WRKFL-2650 - Add Labels to Sidebar Items

kylerberry opened the pull request:
## What this PR does <!-- Is it a bugfix? Feature? Why is this needed? -->
Modify the Sidebar component to support labels.

## Screenshots & Recordings <!-- It's really helpful to give your reviewer some extra context - A picture is worth a thousand words after all. -->
<img width="99" alt="Screenshot 2023-05-30 at 1 29 16 PM" src="https://github.com/vimeo/iris/assets/2123735/33a18544-d71d-477e-88c3-f4fd8a50a195">

**New control added to show/hide labels**
![Screenshot 2023-05-23 at 11 22 03 AM](https://github.com/vimeo/iris/assets/2123735/4e6f9f7b-fe90-430e-9286-26e442d951ed)

## How it does that <!-- implementation details, design decisions, info to help the PR reviewer, etc -->
Adds a `labelAsTooltip` property to inform the sidebar items whether or not to use the provided label as a tooltip or as label text. Modifies some widths here and there to support larger items now that text is included

## Testing <!-- instructions on how to test -->
Check it out in storybook
https://vimeo.github.io/iris/sb/WRKFL-2650/?path=/story/components-sidebar--controls&args=labelAsTooltip:false

Boughtmanatee5 commented on src/components/Sidebar/Sidebar.minors.tsx:
```diff
@@ -31,6 +32,20 @@ export function Item({
     </Tip>
+  ) : (
+    <ItemLabeledStyled
+      aria-label={label}
```
We shouldn't need the aria-label if the text is actually rendered inside the button. A screen reader will use that text. What we might want to do is add `aria-hidden` to the icon so it skips the icon and just reads the next.

kylerberry replied:
agreed. Although I think it may be up to the implementor to add the `aria-hidden` to the icon, the icon props at this juncture are readonly.

kylerberry committed:
WRKFL-2650 - Add Labels to Sidebar Items

Modify the Sidebar component to support labels instead of tooltips.
Adjust the slideout panel to accomodate for a wider sidebar.

kylerberry commented on src/components/Sidebar/Sidebar.tsx:
```diff
@@ -31,6 +36,12 @@ function SidebarComponent({
+  const sidebarRef = useRef<HTMLDivElement | null>(null);
+  const [sidebarWidth, setSidebarWidth] = useState(0);
+
+  useEffect(() => {
```
Since the sidebar now has a variable width to support different length text labels, The panel slideout effect needs an accompanying offset. The best way I could think to do this was to add an offset to the panel informed by the width of the sidebar (instead of a static `4.5rem`).

juliewongbandue replied:
Should `useLayoutEffect` be used here then?

kylerberry replied:
👍

juliewongbandue commented on src/components/Sidebar/Sidebar.minors.tsx:
```diff
@@ -55,6 +69,38 @@ const ItemStyled = styled(Button)<{
+    margin: 0 0 0.45rem 0;
+  }
+
+  > span {
```
I would add a comment here describing what this `span` is here inside the button. Ie: "Button label"

juliewongbandue commented on src/components/Sidebar/Sidebar.story.tsx:
```diff
@@ -21,37 +27,60 @@ const Header = styled(H)`
 `;
 
-const Template: Story<Props> = ({ attach = 'left', ...args }) => {
+const Template: Story<Props & { labelAsTooltip?: boolean }> = ({
```
If you add `labelAsTooltip` to `ItemPropsIntrinsic` you can change this to `Story<Props & Pick<ItemPropsIntrinsic, 'labelAsTooltip'>>` so you can just pass in `args.labelAsTooltip` to the `Sidebar.Item`s below...

juliewongbandue commented on src/components/Sidebar/Sidebar.story.tsx:
```diff
@@ -21,37 +27,60 @@ const Header = styled(H)`
-const Template: Story<Props> = ({ attach = 'left', ...args }) => {
+const Template: Story<Props & { labelAsTooltip?: boolean }> = ({
+  attach = 'left',
+  labelAsTooltip = true,
```
You can also remove this and add `labelAsTooltip: { control: 'boolean' }` so designers and engineers can toggle the labels on and off if using the storybook controls. 👍

juliewongbandue reviewed (changes requested):
Left some comments. Let me know if you have any questions!

kylerberry commented on src/components/Sidebar/Sidebar.minors.tsx:
```diff
@@ -55,6 +69,38 @@ const ItemStyled = styled(Button)<{
+  ${({ isActive, activeStyles }) => (isActive ? activeStyles : '')};
+
+  > svg {
+    padding: 0 !important;
```
no idea, I copied it from the other `Sidebar.Item`

juliewongbandue commented on src/components/Sidebar/Sidebar.minors.tsx:
```diff
@@ -12,12 +12,13 @@ export function Item({
   isActive,
   activeStyles,
   ...props
 }: ItemPropsIntrinsic) {
```
Yep! That's right! 👍

juliewongbandue commented on src/components/Sidebar/Sidebar.story.tsx:
```diff
@@ -14,44 +20,66 @@ export default {
-          <Header size="3">Item 1</Header>
+        <Sidebar.Item
+          label="Privacy"
+          labelAsTooltip={labelAsTooltip}
```
One more thing: you should use the `labelAsTooltip` on the `args` object so the storybook control can work... so instead of just `labelAsTooltip` just change this to `args.labelAsTooltip`

kylerberry commented:
@juliewongbandue going to close for now because there were some deeper suggestions that Seth had. Will need to rework a couple things

kylerberry committed:
update to support a default hover and active state

kylerberry committed:
update the break margin and remove a comment

kylerberry committed:
WRKFL-2650 remove activeStyles changes

kylerberry commented on src/components/Sidebar/Sidebar.tsx:
```diff
@@ -31,6 +36,12 @@ function SidebarComponent({
+  const sidebarRef = useRef<HTMLDivElement | null>(null);
+  const [sidebarWidth, setSidebarWidth] = useState(0);
+
+  useLayoutEffect(() => {
```
Using the layout effect to determine the panel's slide-out offset (so that it isn't obscured by the sidebar)

The sidebar width changes based on whether the items have labels or not. So instead of using static widths, this is automatic.

kylerberry commented on src/components/Sidebar/Sidebar.tsx:
```diff
@@ -65,7 +76,7 @@ function SidebarComponent({
       attach: attach === 'left' ? 'right' : 'left',
       onClick: toggle(child),
       key,
-      isActive: activeStyles && active === child.props.label,
```
I removed `activeStyles` here because it seemed weird to derive the active state from whether `activeStyles` prop was provided. IMO it's active state is separate. The change [here](https://github.com/vimeo/iris/pull/271/files#diff-41f48a2ccb4296a2dd74286e47d6d9d9a60b1bbfbf32768ae012b4afcfcc60ffR59) accounts for this

kylerberry commented on src/components/Sidebar/Sidebar.types.ts:
```diff
@@ -26,6 +26,7 @@ export interface ItemPropsExtrinsic {
   children?: ReactElement;
   icon?: ReactElement;
   label: string;
+  labelAsTooltip?: boolean;
```
this prop is yucky but felt like the most suitable way to maintain backwards compatibility

kylerberry commented on src/components/Sidebar/Sidebar.minors.tsx:
```diff
@@ -42,7 +56,7 @@ const ItemStyled = styled(Button)<{
   padding: 0.25rem 0.25rem 0.25rem 0.75rem;
   justify-content: start;
-  ${({ isActive, activeStyles }) => (isActive ? activeStyles : '')};
+  ${({ isActive, activeStyles }) => (isActive && activeStyles ? activeStyles : '')};
```
related to [this change](https://github.com/vimeo/iris/pull/271/files#r1211924697)

kylerberry commented on src/components/Sidebar/Sidebar.minors.tsx:
```diff
@@ -55,9 +70,52 @@ const ItemStyled = styled(Button)<{
   }
 `;
 
+const ItemLabeledStyled = styled(Button)<{
```
disagree. I speak american english 😆 @juliewongbandue https://www.grammarly.com/blog/labeled-labelled/

juliewongbandue replied:
I just thought we'd keep it consistent with `aria-labelledby`: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby

kylerberry commented on src/components/Sidebar/Sidebar.minors.tsx:
```diff
@@ -55,9 +70,52 @@ const ItemStyled = styled(Button)<{
+    // !important: Overrides padding placed on svg in Stylebar.style.ts
+    padding: 0 !important;
+    width: 1.25rem;
+    min-width: 1.25rem;
+    max-width: 1.25rem;
```
It's actually needed to override the `min-height` and `min-width` on the Icon which are larger than the width we need (and they take precident over `width` and `height`)

When removed the icon is too big and pushes the label out of the button area

kylerberry commented on src/components/Sidebar/Sidebar.minors.tsx:
```diff
@@ -55,9 +70,52 @@ const ItemStyled = styled(Button)<{
+    min-width: 1.25rem;
+    max-width: 1.25rem;
+    height: 1.25rem;
+    min-height: 1.25rem;
+    max-height: 1.25rem;
```
It's actually needed to override the `min-height` and `min-width` on the Button which are larger than the width we need (and they take precident over `width` and `height`)

When removed the button is wider than the expected 64px

kylerberry committed:
WRKFL-2650 update minor name

juliewongbandue reviewed (approved):
Thanks for applying all those changes! 👍
